### PR TITLE
Make output chunk spec optional

### PR DESCRIPTION
**Pull request by kwen2501**
`output_chunk_spec` was used to specify how micro-batch outputs should be merged or reduced (if the output field is a loss).

### Previous
Previously, user would need to specify an `output_chunk_spec` to `PipelineDriver`.
For example:
```
output_chunk_spec = {'loss': CustomReducer(torch.tensor(0.0), lambda a, b: a + b),
                     'logits': TensorChunkSpec(0),
                     'encoder_last_hidden_state': TensorChunkSpec(0)}
```

### Now
We automate this process by:
1. Adding an optional `loss_reducer` argument to `PipelineDriver`, which has a default value of `sum_reducer`, i.e.
```
sum_reducer = CustomReducer(torch.tensor(0.0), lambda a, b: a + b)
```
2. If an output field is specified/identified as loss in the earlier `Pipe.from_tracing` call, we would use the provided `loss_reducer` for this field. If `loss_reducer` is not specified, we would use the default `sum_reducer`.
3. For a non-loss output field, we would use `TensorChunkSpec(0)` by default, i.e. merging along the 0 dimension.

### CODE REVIEW

**Comment by HamidShojanazeri** [comment]
Testing the PR, see its failing on "'generated_loss_spec' referenced before assignment" in IR.py https://gist.github.com/HamidShojanazeri/5db4af143f051624ad42b7bf35b8b626

**Review by HamidShojanazeri — COMMENTED** [review]
Thanks @kwen2501 , testing the PR, see its failing on "'generated_loss_spec' referenced before assignment" in IR.py https://gist.github.com/HamidShojanazeri/5db4af143f051624ad42b7bf35b8b626

**Comment by kwen2501** [comment]
> Thanks @kwen2501 , testing the PR, see its failing on "'generated_loss_spec' referenced before assignment" in IR.py https://gist.github.com/HamidShojanazeri/5db4af143f051624ad42b7bf35b8b626

Thanks. Fixed for inference mode in [39db031](https://github.com/pytorch/tau/pull/742/commits/39db031b7972f09650bc80fa2b44be96649c7450)

**Comment by HamidShojanazeri** [comment]
Thanks @kwen2501 , could successfully test it.
